File: ip_getter.py
from nslookup import Nslookup
from rich.progress import track
from rich.progress import Progress, SpinnerColumn, TextColumn, BarColumn, TaskProgressColumn
from rich.console import Console
from rich.text import Text
import time
import os
import argparse

class Lookup():
    def __init__(self, arg):
        self.dns_query = Nslookup(dns_servers=["9.9.9.9", "194.242.2.2"], verbose=False, tcp=False) # uses Quad9 and mullvad dns
        if arg == "None":
            self.default_files = [
                "advertisement.txt",
                "csam.txt",
                "fingerprinting.txt",
                "forums.txt",
                "malware.txt",
                "porn.txt",
                "spam.txt",
                "suspicious.txt",
                "telemetry.txt",
                "to_monitor.txt",
                "tracking.txt"
            ]
        else:
            self.default_files = [
                arg
            ]
        self.ip_files = self.get_ip_files()
        self.domain_list_file_content = []
        self.ip_list_file_content = []
        self.console = Console()
        
        
    def get_ip_files(self):
        ip_files = []
        for file_name in self.default_files:
            ip_files.append(file_name.replace(".txt", ".ip"))
        
        return ip_files
    
    def get_file_content(self, domain_list_file, ip_file):
        content_in_dir = os.listdir(os.getcwd())
        if domain_list_file not in content_in_dir:
            with open(domain_list_file, mode='a'): pass
        else:
            with open(domain_list_file, "r") as ftxt:
                self.domain_list_file_content = ftxt.readlines()
        
        if ip_file not in content_in_dir:
            with open(ip_file, mode='a'): pass
        else:
            with open(ip_file, "r") as fip:
                self.ip_list_file_content = fip.readlines()
    
    def dns_lookup(self, domain):
        ips_record = self.dns_query.dns_lookup(domain.strip())
        return ips_record.answer


    def run(self):
        with Progress(
            SpinnerColumn(),
            TextColumn("[progress.description]{task.description}"),
            BarColumn(),
            TaskProgressColumn(),
            console=self.console
        ) as progress:
            
            file_task = progress.add_task("Processing files", total=len(self.default_files))
            
            for file_name in self.default_files:
                self.console.print(f"[bold cyan]Processing {file_name}[/bold cyan]")
                index = self.default_files.index(file_name)
                ip_file_name = self.ip_files[index]
                self.get_file_content(file_name, ip_file_name)
                domain_task = progress.add_task("", total=len(self.domain_list_file_content))

                progress.update(domain_task, description=f"Processing {file_name}", completed=0, total=len(self.domain_list_file_content))
                
                with open(ip_file_name, "a") as f:
                    for domain in self.domain_list_file_content:
                        if not domain.startswith("#") and not domain == "\n":
                            ips = self.dns_lookup(domain)
                            if len(ips) == 0:
                                self.console.print(f"[bold red][-] no IPs found for {domain.strip()}[/bold red]")
                                
                            else:
                                
                                count_ip_to_add = 0
                                for ip in ips:
                                    if f"{ip}\n" not in self.ip_list_file_content or f"{ip}" not in self.ip_list_file_content:
                                        count_ip_to_add += 1
                                
                                if count_ip_to_add > 0:
                                    f.write(f"\n# {domain}")
                                    self.ip_list_file_content.append(f"\n# {domain}")

                                    for ip in ips:
                                        # if ip == "3.251.41.252": # debug
                                        #     print("test")
                                        if f"{ip}\n" not in self.ip_list_file_content:
                                            self.ip_list_file_content.append(ip)
                                            f.write(f"{ip}\n")

                                            domain_print = f"[bold green]{domain.strip()}[/bold green]"
                                            self.console.print(f"[bold cyan][+][/bold cyan] {ip} for {domain_print} added to {ip_file_name}")
                                    
                                    ips.clear()
                                
                                else:
                                    
                                    self.console.print(f"[bold yellow][*] no new IPs for {domain.strip()}[/bold yellow]")
                        
                        time.sleep(0.0001)

                        progress.advance(domain_task)
                
                progress.advance(file_task)

                time.sleep(1)

parser = argparse.ArgumentParser(description='Your script description')
parser.add_argument('-f', '--file', type=str, help='Name of the file to process')
args = parser.parse_args()
if args.file:
    lookup = Lookup(arg=args.file)
else:
    lookup = Lookup(arg="None")

lookup.run()

File: ip_getter.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"strings"
	"sync"

	"github.com/miekg/dns"
	"github.com/schollz/progressbar/v3"
)

type Lookup struct {
	defaultFiles          []string
	ipFiles               []string
	domainListFileContent []string
	ipListFileContent     []string
}

func NewLookup(files []string) *Lookup {
	return &Lookup{
		defaultFiles: files,
			ipFiles:      buildIPFiles(files),
	}
}

func buildIPFiles(files []string) []string {
	var res []string
	for _, f := range files {
		res = append(res, strings.TrimSuffix(f, ".txt")+".ip")
	}
	return res
}

func (l *Lookup) loadFileContent(domainFile, ipFile string) error {
	touch := func(path string) {
		if _, err := os.Stat(path); os.IsNotExist(err) {
			os.WriteFile(path, []byte{}, 0644)
		}
	}
	touch(domainFile)
	touch(ipFile)
	domainData, err := os.ReadFile(domainFile)
	if err != nil {
		return err
	}
	l.domainListFileContent = strings.Split(string(domainData), "\n")
	ipData, err := os.ReadFile(ipFile)
	if err != nil {
		return err
	}
	l.ipListFileContent = strings.Split(string(ipData), "\n")
	return nil
}

func dnsLookup(domain string, dnsServers []string) ([]string, error) {
	var ips []string
	c := new(dns.Client)
	for _, server := range dnsServers {
		m := new(dns.Msg)
		m.SetQuestion(dns.Fqdn(domain), dns.TypeA)
		r, _, err := c.Exchange(m, server+":53")
		if err != nil {
			continue
		}
		if r.Rcode != dns.RcodeSuccess {
			continue
		}
		for _, ans := range r.Answer {
			if a, ok := ans.(*dns.A); ok {
				ips = append(ips, a.A.String())
			}
		}
	}
	return ips, nil
}

func contains(list []string, item string) bool {
	for _, x := range list {
		if strings.TrimSpace(x) == strings.TrimSpace(item) {
			return true
		}
	}
	return false
}

func (l *Lookup) run(workers int) {
	fileBar := progressbar.NewOptions(len(l.defaultFiles),
					  progressbar.OptionSetDescription("Processing files"),
					  progressbar.OptionSetPredictTime(false),
					  progressbar.OptionShowCount(),
	)
	dnsServers := []string{"9.9.9.9", "9.9.9.10", "194.242.2.2"}
	for i, fileName := range l.defaultFiles {
		fmt.Printf("\n[ Processing %s ]\n", fileName)
		ipFile := l.ipFiles[i]
		l.loadFileContent(fileName, ipFile)
		domainBar := progressbar.NewOptions(len(l.domainListFileContent),
						    progressbar.OptionSetDescription(fmt.Sprintf("Processing %s", fileName)),
						    progressbar.OptionSetPredictTime(false),
						    progressbar.OptionShowCount(),
		)
		f, err := os.OpenFile(ipFile, os.O_APPEND|os.O_WRONLY, 0644)
		if err != nil {
			fmt.Println("Error creating ip file:", err)
			continue
		}
		writer := bufio.NewWriter(f)
		var wg sync.WaitGroup
		sem := make(chan struct{}, workers)
		var mu sync.Mutex
		for _, domain := range l.domainListFileContent {
			domain = strings.TrimSpace(domain)
			if domain == "" || strings.HasPrefix(domain, "#") {
				domainBar.Add(1)
				continue
			}
			if idx := strings.Index(domain, "#"); idx != -1 {
				domain = strings.TrimSpace(domain[:idx])
			}
			if domain == "" {
				domainBar.Add(1)
				continue
			}
			wg.Add(1)
			go func(domain string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()
				ips, _ := dnsLookup(domain, dnsServers)
				if len(ips) == 0 {
					fmt.Printf("[ - ] No IPs found for %s\n", domain)
					domainBar.Add(1)
					return
				}
				mu.Lock()
				defer mu.Unlock()
				newIPs := 0
				for _, ip := range ips {
					if !contains(l.ipListFileContent, ip) {
						newIPs++
					}
				}
				if newIPs > 0 {
					writer.WriteString("\n# " + domain + "\n")
					l.ipListFileContent = append(l.ipListFileContent, "# "+domain)
					for _, ip := range ips {
						if !contains(l.ipListFileContent, ip) {
							writer.WriteString(ip + "\n")
							l.ipListFileContent = append(l.ipListFileContent, ip)
							fmt.Printf("[ + ] %s for %s added to %s\n", ip, domain, ipFile)
						}
					}
				} else {
					fmt.Printf("[ * ] No new IPs for %s\n", domain)
				}
				domainBar.Add(1)
			}(domain)
		}
		wg.Wait()
		writer.Flush()
		f.Close()
		fileBar.Add(1)
	}
}

func main() {
	filePtr := flag.String("f", "None", "Name of the file(s) to process (space-separated, or 'None' for default list)")
	workersPtr := flag.Int("w", 20, "Number of workers (goroutines) to use")
	flag.Parse()

	var files []string
	if *filePtr == "None" {
		files = []string{
			"advertisement.txt",
			"fingerprinting.txt",
			"forums.txt",
			"malware.txt",
			"spam.txt",
			"suspicious.txt",
			"telemetry.txt",
			"to_monitor.txt",
			"tracking.txt",
			"zoophilia.txt",
			"porn.txt",
		}
	} else {
		files = strings.Fields(*filePtr)
	}

	lookup := NewLookup(files)
	lookup.run(*workersPtr)
}
